addons/l10n_in/models/product_template.py

```diff
@@ -2,21 +2,28 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 from odoo import api, models, fields, _
-from odoo.exceptions import ValidationError
 
 
 class ProductTemplate(models.Model):
     _inherit = 'product.template'
 
     l10n_in_hsn_code = fields.Char(string="HSN/SAC Code", help="Harmonized System Nomenclature/Services Accounting Code")
     l10n_in_hsn_description = fields.Char(string="HSN/SAC Description", help="HSN/SAC description is required if HSN/SAC code is not provided.")
+    l10n_in_hsn_warning = fields.Text(string="HSC/SAC warning", compute="_compute_l10n_in_hsn_warning")
 
-    @api.constrains('l10n_in_hsn_code')
-    def _check_hsn_code_validation(self):
+    @api.depends('sale_ok', 'l10n_in_hsn_code')
+    def _compute_l10n_in_hsn_warning(self):
+        digit_suffixes = {
+                '4': _("either 4, 6 or 8"),
+                '6': _("either 6 or 8"),
+                '8': _("8")
+            }
+        active_hsn_code_digit_len = self.env.companies.mapped('l10n_in_hsn_code_digit')
+        # Remove False value if any company have False value in l10n_in_hsn_code_digit
+        active_hsn_code_digit_len = list(filter(bool, active_hsn_code_digit_len))
```

Comment:
Just compute the maximum here instead of doing it in the loop below
```suggestion
    active_hsn_code_digit_len = max(
        int(company.l10n_in_hsn_code_digit) 
        for company in self.env.companies
    )
```

Note that it works also if the field is not set because `int(False) == 0`